src/main/java/co/thatch/example/metrics/InMemoryMetricsService.java

```diff
@@ -0,0 +1,117 @@
+package co.thatch.example.metrics;
+
+import co.thatch.example.metrics.model.Metric;
+import co.thatch.example.metrics.model.MetricBucket;
+import io.micronaut.runtime.server.event.ServerStartupEvent;
+import io.micronaut.runtime.event.annotation.EventListener;
+import io.micronaut.scheduling.annotation.Async;
+
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.DoubleAdder;
+import java.util.stream.LongStream;
+
+
+@Singleton
+public class InMemoryMetricsService implements MetricsService {
+    private static final int MAX_WINDOW_SIZE = 1_800;
+    private static final int MAX_WINDOW_SIZE_WITH_BUFFER = MAX_WINDOW_SIZE + 1;
+
+    private final DoubleAdder requestCounter = new DoubleAdder();
+    private final double[] requestCountRecordings = new double[MAX_WINDOW_SIZE_WITH_BUFFER];
+    private final HashMap<Long, MetricHolder> metrics = new HashMap<>();
+    private final PriorityQueue<Long> timestamps = new PriorityQueue<>();
+
+    @Async
+    @EventListener
+    public void onStartup(ServerStartupEvent event) throws InterruptedException {
+        while (true) {
+            requestCountRecordings[Math.toIntExact(Instant.now().getEpochSecond() % requestCountRecordings.length)] = requestCounter.doubleValue();
```

Comment:
> Math.toIntExact(Instant.now().getEpochSecond()

This will make testing a little harder since you don't have so much controll over the flow of time. How could you make this more test/mock friendly?